java-checks/src/main/java/org/sonar/java/checks/AssertTrueInsteadOfDedicatedAssertCheck.java

```diff
@@ -0,0 +1,205 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.NULL_LITERAL;
+
+@Rule(key = "S5785")
+public class AssertTrueInsteadOfDedicatedAssertCheck extends AbstractMethodDetection {
+
+  private static final String[] ASSERT_METHOD_NAMES = {"assertTrue", "assertFalse"};
+  private static final String[] ASSERTION_CLASSES = {
+    // JUnit4
+    "org.junit.Assert",
+    "junit.framework.TestCase",
+    // JUnit4 (deprecated)
+    "junit.framework.Assert",
+    // JUnit5
+    "org.junit.jupiter.api.Assertions"
+  };
+
+  private enum Assertion {
+    NULL("Null", "A null-check"),
+    NOT_NULL("NotNull", "A null-check"),
+    SAME("Same", "An object reference comparison"),
+    NOT_SAME("NotSame", "An object reference comparison"),
+    EQUALS("Equals", "An equals check"),
+    NOT_EQUALS("NotEquals", "An equals check");
+
+    public final String methodName;
+    public final String actionDescription;
+
+    Assertion(String namePostfix, String actionDescription) {
+      methodName = "assert" + namePostfix;
+      this.actionDescription = actionDescription;
+    }
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.create().ofTypes(ASSERTION_CLASSES).names(ASSERT_METHOD_NAMES).withAnyParameters().build();
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    Arguments arguments = mit.arguments();
+
+    ExpressionTree argumentExpression;
+    if (hasBooleanArgumentAtPosition(arguments, 0)) {
+      argumentExpression = arguments.get(0);
+    } else if (hasBooleanArgumentAtPosition(arguments, arguments.size() - 1)) {
+      argumentExpression = arguments.get(arguments.size() - 1);
+    } else {
+      // We encountered a JUnit5 assert[True|False] method that accepts a BooleanSupplier - not supported.
+      return;
```

Comment:
Let's add a test case to cover this case, and we can even use a false-negative to explicitly show the limitation of this rule:
```
Assertions.assertTrue(() -> foo == bar); // false-negative because BooleanSupplier is not supported by this rule
```